pandas/tools/hashing.py

```diff
@@ -80,53 +80,56 @@ def hash_array(vals, encoding='utf8', hash_key=None):
     encoding : string, default 'utf8'
         encoding for data & key when strings
     hash_key : string key to encode, default to _default_hash_key
+    categorize : bool, default True
+        Whether to first categorize object arrays before hashing. This is more
+        efficient when the array contains duplicate values.
 
     Returns
     -------
     1d uint64 numpy array of hash values, same length as the vals
 
     """
 
-    # work with cagegoricals as ints. (This check is above the complex
-    # check so that we don't ask numpy if categorical is a subdtype of
-    # complex, as it will choke.
     if hash_key is None:
         hash_key = _default_hash_key
 
+    # For categoricals, we hash the categories, then remap the codes to the
+    # hash values. (This check is above the complex check so that we don't ask
+    # numpy if categorical is a subdtype of complex, as it will choke.
     if is_categorical_dtype(vals.dtype):
-        vals = vals.codes
+        cat_hashed = hash_array(vals.categories.values, encoding, hash_key,
```

Comment:
Can you explain where you mean "like we do below"? I'm assuming you mean lines 125-130, where we handle object dtype.

For categoricals, what we want to do is:
- Get a hash for the `categories`. This should match what `hash_pandas_object(series, index=False)` would return for the un-categorized data. Meaning `hash_pandas_object(object_series) == hash_pandas_object(object_series.astype('category'))`.
- Remap the category hashes based on the codes.

`vals.categories.values` can be of any dtype, so we need to recurse through `hash_array` again to get the hashes of the categories. However, we also know that the categories are already unique, so we don't want to call `factorize` again. As such, we set `categorize=False` to skip that. We can't just call `hash_object_array`, as the categories may not be objects. And we need to do the remapping, so we can't just set `vals = something` and fall through like we did before.

I don't think adding this extra keyword overly complicates things, and do think this is the simplest way to do this. I may not be understanding what you're trying to suggest here - perhaps if you could explain a bit better I might get it.